### PR TITLE
Update to current upstream PDCLib

### PR DESCRIPTION
**Pull request by thrimbor**
Major improvements are:
- atexit() was fixed (this fixes destructors of global objects in C++)
- fseek() fix
- Usage of upstream `threads.h`
- threadsafe stream I/O
- Usage of dlmalloc as our allocator (this fixes the SDL_ttf sample hang on complex_4627debug)

To see the changes I did to dlmalloc to make it fit, diff it with the one in the example platform.

All nxdk samples were run in xqemu to test this PR, further testing is appreciated.

This should be merged without rebase to ensure the submodule references in nxdk stay intact.

### CODE REVIEW

**Comment by JayFoxRox** [comment]
> Usage of dlmalloc as our allocator (this fixes the SDL_ttf sample hang on complex_4627debug)

Do we have an issue for this? Why did this happen?
As we had no allocator, this sounds like a bug in the sample, that still has to be fixed independently?

File: platform/xbox/crt0.c [hunk 1]
```diff
@@ -6,15 +6,32 @@
...
+extern struct _PDCLIB_file_t * stderr;
 
 static int main_wrapper ()
 {
     _PDCLIB_xbox_tss_init();
 
+    mtx_init(&_PDCLIB_filelist_mtx, mtx_plain);
+    mtx_init(&stdin->mtx, mtx_recursive);
+    mtx_init(&stdout->mtx, mtx_recursive);
+    mtx_init(&stderr->mtx, mtx_recursive);
```
Comment (JayFoxRox) [inline platform/xbox/crt0.c:19]: nit: I wonder if this should be moved into a separate function, for users who don't use crt0 (my DXT code for example; but also some of my other lower-level project).

**Comment by thrimbor** [comment]
> > Usage of dlmalloc as our allocator (this fixes the SDL_ttf sample hang on complex_4627debug)
> 
> Do we have an issue for this? Why did this happen?
> As we had no allocator, this sounds like a bug in the sample, that still has to be fixed independently?

When I encountered this I traced it back to PDCLib's malloc() implementation. During one of the allocations done by SDL_ttf, it would try to split a block of memory with invalid size (the size was larger than Xbox RAM, although I don't remember the exact size) and would try to write to unmapped memory while writing the header of the newly created block. I suspected it to be a bug in PDCLib, but didn't get around to nailing it down.

File: platform/xbox/crt0.c (hunk 1 above)
Comment (thrimbor) [reply platform/xbox/crt0.c:19]: I pushed an update that does this.

**Comment by JayFoxRox** [comment]
If that old malloc code still exists in pdclib, we should still create an issue upstream (others might use it / we might use it in special cases), linking to respective nxdk and pdclib revisions.

File: platform/xbox/crt0.c [hunk 2]
```diff
@@ -6,15 +6,42 @@
...
+    mtx_destroy(&stdin->mtx);
+    mtx_destroy(&stdout->mtx);
+    mtx_destroy(&stderr->mtx);
+}
 
 static int main_wrapper ()
 {
     _PDCLIB_xbox_tss_init();
 
+    _PDCLIB_xbox_run_libc_initializers();
```
Comment (JayFoxRox) [inline platform/xbox/crt0.c]: nit: as this doesn't run a list of callbacks, I feel that `_PDCLIB_xbox_libc_init` is more suitable (closer to the `tss_init`)

Comment (thrimbor) [reply platform/xbox/crt0.c]: Agreed and changed.

**Comment by thrimbor** [comment]
> If that old malloc code still exists in pdclib, we should still create an issue upstream (others might use it / we might use it in special cases), linking to respective nxdk and pdclib revisions.

The old malloc code was removed when dlmalloc was introduced. I still plan to make sure that this was not an issue of our glue code or some issue with SDL_ttf when I get around to it.

**Comment by JayFoxRox** [comment]
Looks (almost) good to me.

Why is there a commit named "Merge branch 'nxdk' into nxdk_next" ?

Also I assume this breaks the history in some commits, when upstream was updated, but our xbox-port wasn't. However, it's probably tricky to integrate this properly with how the upstream merging works (which also doesn't always pay attention to this).

I'm also still not happy about how `dlmalloc` is integrated, but that's an upstream issue (adds to my frustration with pdclib in certain areas).

> This should be merged without rebase to ensure the submodule references in nxdk stay intact.

This should't be an issue? GitHub keeps the commits - it just hides them from the history.
Anyhow, fine with merging, too.

However, GitHub does't allow us to rebase-and-merge either: "This branch cannot be rebased due to conflicts" - what's up with that?

**Comment by thrimbor** [comment]
> Why is there a commit named "Merge branch 'nxdk' into nxdk_next" ?

I merged the nxdk-branch into a branch containing the upstream master. It produces a history that is easier to understand in tools like git gui (compared to merging the other way)
.
> Also I assume this breaks the history in some commits, when upstream was updated, but our xbox-port wasn't. However, it's probably tricky to integrate this properly with how the upstream merging works (which also doesn't always pay attention to this).

Some commits will not build with nxdk, yes. I know no way around this, though (apart from rebasing, which has other problems).

> > This should be merged without rebase to ensure the submodule references in nxdk stay intact.
> 
> This should't be an issue? GitHub keeps the commits - it just hides them from the history.

GitHub blocks access to unreferenced commits (I recall mentioning this on Discord), which, if I remember correctly, has already been an issue with our SDL submodule in the past. Additionally, if GitHub follows standard git behavior, unreachable commits will get garbage collected eventually.
We could get around this though by creating a branch or a tag that references the old HEAD.

> However, GitHub does't allow us to rebase-and-merge either: "This branch cannot be rebased due to conflicts" - what's up with that?

GitHub would try to remove the merge commits while rebasing, which would lead to multiple commits containing the same change (so the conflicts are caused by our downstream changes, which appear both in the merge and our current master).

**Comment by JayFoxRox** [comment]
Please always leave a comment as to why a PR or issue was closed (accidental or otherwise)

**Pull request closed without merging**